src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetServiceImpl.java

```diff
@@ -38,15 +41,27 @@ public String createDataset(String datasetName, Country country, Language langua
     checkArgument(!records.isEmpty(), "Records must not be empty");
 
     var entity = new DatasetEntity(datasetName, records.size());
-    String id;
     try {
-      id = datasetRepository.save(entity).getDatasetId();
+      entity = datasetRepository.save(entity);
     } catch (Exception e) {
       throw new ServiceException("Error creating the dataset: " + e.getMessage(), e);
     }
 
-    var dataset = generatorService.generate(id, datasetName, country, language, records);
+    var dataset = generatorService
+        .generate(entity.getDatasetId(), datasetName, country, language, records);
+
+    // if there are duplicate records in the original list
+    if (dataset.getRecords().size() < records.size()) {
+      // adjust records qty to be equal to non duplicate records
+      entity.setRecordsQuantity(dataset.getRecords().size());
+      try {
+        datasetRepository.save(entity);
+      } catch (Exception e) {
+        throw new ServiceException("Error updating the dataset: " + e.getMessage(), e);
+      }
+    }
```

Comment:
Unless I'm mistaken, I think you can do this check before saving the entity the first time. I think that would be better... first create the entity completely with all counters and then save it once?

Reply:
In line 50 I need the datasetId to call generatorService, which is in charge of creating the Set<Records> (by reading the List< ByteArrayInputStream>). Each Record object contains the datasetId, then after calling generatorService I know how many records are duplicates, that is why I update the entity here, but I was not able to produce all counters at first.

Reply:
You're right, I missed that. But still there must be a way... but perhaps that's not the most important thing to work on right now. This only occurs when there are duplicates, which should not happen too often.

Reply:
Yes I thought it should not be often,  but I will create a story to not forget about this

Reply:
[MET-2760](https://europeana.atlassian.net/browse/MET-2760)